pkg/operator/configobservation/apiserver/observe_admission_plugins.go

```diff
@@ -0,0 +1,95 @@
+package apiserver
+
+import (
+	"fmt"
+	"slices"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type pluginCheckerFunc func(listers configobservation.Listers) (enabled, disabled []string, err error)
+
+var (
+	enableAdmissionPluginsPath  = []string{"apiServerArguments", "enable-admission-plugins"}
+	disableAdmissionPluginsPath = []string{"apiServerArguments", "disable-admission-plugins"}
+
+	pluginCheckers = []pluginCheckerFunc{
+		roleBindingRestrictionPluginChecker,
+	}
+)
+
+// ObserveAdmissionPlugins manages the apiServerArguments.enable-admission-plugins and
+// apiServerArguments.disable-admission-plugins fields of the configuration. It defines a list of
+// plugin checkers which check the state of specific plugins, and add them to the enabled or disabled
+// list as required. This observer will overwrite any pre-existing values of the two fields in the existingConfig.
+func ObserveAdmissionPlugins(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]any) (ret map[string]any, _ []error) {
+	defer func() {
+		ret = configobserver.Pruned(ret, enableAdmissionPluginsPath, disableAdmissionPluginsPath)
+	}()
+
+	if len(pluginCheckers) == 0 {
+		return existingConfig, nil
+	}
+
+	listers := genericListers.(configobservation.Listers)
+	enabledSet := sets.New[string]()
+	disabledSet := sets.New[string]()
+
+	for _, pluginChecker := range pluginCheckers {
+		enabled, disabled, err := pluginChecker(listers)
+		if err != nil {
+			return existingConfig, []error{err}
+		}
+
+		enabledSet.Insert(enabled...)
+		disabledSet.Insert(disabled...)
+	}
+
+	if intersection := enabledSet.Intersection(disabledSet); intersection.Len() > 0 {
+		return existingConfig, []error{fmt.Errorf("plugins cannot be enabled and disabled at the same time: %v", intersection.UnsortedList())}
+	}
+
+	observedConfig := map[string]any{}
+
+	if enabledSet.Len() > 0 {
```

Comment:
if the length of the enabled set is 0, should this ensure that the returned configuration is that this nested slice is actually empty? Is that the same as what just not setting it does? Should there be a test for the case where a previous config observer may have set this and this overrides it to empty when nothing is returned as being in the enabled set?

Reply:
I aimed to preserve a similar behavior as the existing setup; currently, for example, the `--disable-admission-plugins` field is not set in the default config or overrides, and this results in the field missing from the merged configuration. I wrote this config observer in a similar way; if there's no plugins to enable/disable, don't set the flag at all.